---
title: "Spaniel"
author: "Rachel Queen"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{Using Spaniel}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, include = TRUE,
                        fig.height = 8, fig.width = 8, fig.align = "center",
                        echo=TRUE
                        )

```  

## About Spaniel

Spaniel is an R package designed to visualise results of Spatial Transcriptomics
experiments. The current stable version of Spaniel (version 1.1.0) is available from Bioconductor:

```{r install, eval = FALSE}

devtools::install_github("RachelQueen1/Spaniel", repos = "Dev" )


```

### Spaniel - with 10X import option

This vignette refers to a development version of Spaniel (version 1.2) designed to import data from a 10X Genomics Visium experiment.  
This version will be tested and pushed to Bioconductor early in the new year (2020). If you would like to test the features described in this vignette you can install Spaniel using the following command:


```{r install, eval = FALSE}

devtools::install_github("RachelQueen1/Spaniel", repos = "Dev" )


```



```{r, load libraries}
library(Spaniel)
library(DropletUtils)
```

## Data

This vignette will show how to load the results of 10X Visium spatial 
transcriptomics experiment which has been run through the Space Ranger pipeline.
The data is distributed as part of the Space Ranger software package which can 
be downloaded here:

(add in link)

The output from from the ?? is used as an example. 


### Import the expression data

Spaniel requires a data frame of counts where the genes equate to the rows and 
the columns equate to a Spatial Transcriptomics spot/barcode. This can be loaded 
in from SpaceRanger output using the createVisiumSCE function which imports the 
gene expression data, spatial barcodes, and image dimensions into 
the SingleCellExperiment object. 

```{r counts}

sce <- createVisiumSCE(tenXDir="C:/Users/nrq2/Documents/GitHub/outs", 
                            resolution="Low")

```

### SCE Object

The pixel coordinates are added to the colData of the SCE object shown below:


```{r barcodes}
colData(sce)[, c("Barcode", "pixel_x", "pixel_y")]
```

The image dimensions are added to the metadata of the SCE object:

```{r image_dimensions}
metadata(sce)$ImgDims
```

The image is stored as a rasterised grob.

```{r grob}
metadata(sce)$Grob
```

## Quality Control

Assessing the number of genes and number of counts per spot is a useful quality 
control step. Spaniel allows QC metrics to be viewed on top of the 
histological image so that any quality issues can be pinpointed. 
Spots within the tissue region which have a low number of genes or counts may 
be due to experimental problems which should be addressed. Conversely spots 
which lie  outside of the tissue and have a high number of counts or large 
number of genes may indicate that there is background contamination. 


### Visualisation

The plotting function allows the use of a binary filter to visualise which 
spots pass filtering thresholds. We create a filter to show spots at 1 gene is detected. Spots where no genes are detected will be removed from the remainder of the analysis.

__NOTE:__ The parameters are set for subset of counts used in this dataset.  
The filter thresholds will be experiment specific and should be adjusted as 
necessary.

```{r, qcplotting,  results = "hide" }

filter <- sce$detected > 0
spanielPlot(object = sce,
        plotType = "NoGenes", 
        showFilter = filter, 
        techType = "Visium", 
        ptSizeMax = 2)

```


Spots where no genes are detected can be removed from the remainder of the analysis.

```{r}
sce <- sce[, filter]
```

The filtered data can then be normalised using the the "normalize" function from scater and the expression of selected genes can be viewed on the histological image.

```{r, gene plot,  results = "hide" }


sce <- logNormCounts(sce)

gene <- "ENSMUSG00000024843"
p2 <- spanielPlot(object = sce,
        plotType = "Gene", gene = gene,
        showFilter = NULL, 
        techType = "Visium", 
        ptSizeMax = 2)

```



## ShinySpaniel

## Using the RunShinySpaniel function in R

Spaniel includes a shiny app designed to share preprocessed data. The app is 
included as the runShinySpaniel function. 

### Mark Cluster Columns

The shiny Spaniel app has the option of visualising any cluster data on a plot.
First the columns containing the different clustering solutions need to be 
marked in the meta.dat with the prefix "cluster_"  so that Spaniel knows where 
to look for them. 

```{r, markClusters}
seuratFiltered <- markClusterCol(seuratFiltered, "res")
```

### Save data object

```{r, eval = FALSE,  echo=TRUE}
saveRDS(seuratFiltered, "data.rds")
```

An example of processed data set can be found:

```{r, eval = FALSE, echo=TRUE}
file.path(system.file(package = "Spaniel"), "extdata/SeuratData.rds" )
```

### Save image object

```{r, eval = FALSE, echo=TRUE}
saveRDS(image, "image.rds")

```
An example of processed image file  can be found:

```{r, eval = FALSE}
file.path(system.file(package = "Spaniel"), "extdata/image.rds" )
```

### Run shinySpaniel!
To launch the shiny Spaniel app issue the command:
```{r, markclusterCols, eval = FALSE}
runShinySpaniel()

```

## Deploying ShinySpaniel
For sharing analysis it is also possible to deploy the ShinySpaniel app to 
Shiny.io or any other server running R Server. The standalone app is located:

```{r, eval = FALSE}
spanielApp <- file.path(system.file(package = "Spaniel"), "ShinySpaniel" )
```

You can sign up for a shinyapps.io account:

https://www.shinyapps.io/

Sign into your account and follow steps 1 and and 2 in the getting started 
instructions to authorise your account in R. 

Then run the following code:


```{r, eval = FALSE}
library(rsconnect)
rsconnect::deployApp(spanielApp)

```






