---
title: "Spaniel"
author: "Rachel Queen"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## About Spaniel

Spaniel is an R package designed to visualise results of Spatial Transcriptomics
experiments. To install and load Spaniel:


```{r install, include = TRUE, message=FALSE, warning=FALSE}

if (!require("devtools")) 
    install.packages("devtools")
if (!require("Seurat")) 
    devtools::install_github(repo = 'satijalab/seurat', ref = 'release/3.0')
if (!require("Spaniel")) 
    devtools::install_github(repo = "RachelQueen1/Spaniel", build_vignettes = TRUE)

library(Spaniel)

```

## Load data

### Expression Matrix

Spaniel requires a data frame of counts where the genes equate to the rows and 
the columns equate to a Spatial Transcriptomics spot/barcode.

```{r counts, include = TRUE}

### read in test data
counts <- readRDS(file.path(system.file(package = "Spaniel"), 
                            "data/counts.rds"))

```

The barcodes are named in the colnames of the data frame:
```{r colnames, include = TRUE}
colnames(counts)[1:10]
```

The genes are named in rownames of the data frame
```{r rownames, include = TRUE}
rownames(counts)[1:10]
```

### Barcode Coordinates

Spaniel also requires the coordinates of the barcodes. In this example we will 
load the undjusted coordinates provided by spatial transcriptomics but adjusted 
coordinates can also be used. The barcodes need to be stored in a tab delimited
text file. The first column of the file should contain the barcodes, the second
coumn the x coordinate of the barcode and the third column the y 
coordinate.

```{r barcodes, include = TRUE}

barcodesFile <- file.path(system.file(package = "Spaniel"), 
                            "1000L2_barcodes.txt")
```

It is not necessary to load this file in at this point but for demonstration
purposes we will load it now and look at the first few lines of the file.

```{r barcodesTop, include = TRUE}
barcodes <- read.csv(barcodesFile, sep = "\t", header = F)
head(barcodes)
```

## Create a S4 object

Spaniel includes two functions to create S4 objects containing the counts and 
barcode information. These are the Seurat object:
https://satijalab.org/seurat/
and a SingleCellExperiment object: https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html

### Creating a Seurat object

The readSeurat() function can be used to create a Seurat object. The count data 
is stored in the counts slot of the assay slot of the object, the barcodes are 
stored in the meta.data slot and the ProjectName and SectionNumber arguments can 
be used to add information about the Sample and position on slide to the 
project.name slot of the Seurat object.

```{r readSeurat, include = TRUE}

SeuratObj = readSeurat(counts,
                       barcodesFile, 
                       ProjectName = "TestProj", 
                       SectionNumber = 1
                       )

```

The barcodes are stored in the metadata:
```{r readSeurat_meta, include = TRUE}
head(SeuratObj@meta.data)
```

Corner of the count data:
```{r readSeurat_counts, include = TRUE}
SeuratObj@assays$RNA@counts[1:10, 1:5]
```

Project name
```{r readSeurat_project, include = TRUE}
SeuratObj@project.name
```

### Creating a SingleCellExperiment object
Alternatively, the readSCE function can be used to create a SingleCellExperiment
object. This time the ProjectName and SectionNumber are added as a column of the
colData of the object. The number of genes and counts per spot are calculated by 
the calculateQCMetrics() from the scater bioconductor package.

```{r readSCE, include = TRUE, message= F, warning=F}
sce = readSCE(Counts = counts,
              BarcodeFile = barcodesFile, 
              ProjectName = "TestProj", 
              SectionNumber = 1)
```

Barcodes are stored in the colData
```{r readSCE_ColData, include = TRUE}
head(colData(sce)[1:5,1:5])
```

Corner of the count data:
```{r readSCE_counts, include = TRUE}
counts(sce)[1:10, 1:5]
```

Project name:
```{r readSCE_Project, include = TRUE}
sce@colData$project[1]
```

## Load image
The next step is to load the histological image into R. The image example has 
been scaled down in size, using photoshop,  to increase the plotting speeds. 
The image is cropped to around te edges of the spots. 

```{r, fig.show='hold'}
### Load histological image into R

imgFile <- file.path(system.file(package = "Spaniel"), 
                            "HE_Rep1_resized.jpg")

image <- parseImage(imgFile)


```

## Spatial Transcriptomics using Seurat

The rest of this vignette will focus on using a Seurat workflow. 


## Quality Control

Assessing the number of genes and number of counts per spot is a useful quality 
control step. Spaniel allows QC metrics to be viewed on top of the 
histological image so that any quality issues can be pinpointed. 
Spots within the tissue region which have a low number of genes or counts may be 
due to experimental problems which should be addressed. Conversely spots which 
lie  outside of the tissue and have a high number of counts or large number of 
genes may indicate that there is background contamination. 


### Visualisation

The plotting function allows the use of a binary filter to visualise which spots 
pass filtering thresholds. We create a filter to show spots which have greater 
than 2000 genes and a minimum of 300000. These parameters will be experiment 
specific and should be adjusted as necessary.

```{r, qcplotting,  fig.height = 8, fig.width = 8, fig.align = "center", warning= F, message = F}
minGenes = 2000
minUMI = 300000

filter = SeuratObj@meta.data$nFeature_RNA > minGenes &  
    SeuratObj@meta.data$nCount_RNA > minUMI

ST_plot(Object = SeuratObj, 
        Grob = image, 
        PlotType = "NoGenes", 
        ShowFilter = filter)

```

### Filtering

Spots which don't pass QC thresholds can then be filtered from the data:

```{r, filter_seurat }
SeuratFiltered <- subset(x = SeuratObj, subset = nFeature_RNA > minGenes &
                             nCount_RNA > minUMI)

```

## Cluster Spots

A clustering analysis can now be performed using methods found in Seurat 
package. For more details see:

https://satijalab.org/seurat/get_started.html


```{r, find_clusters, message=F, warning=F }



SeuratFiltered <- NormalizeData(object = SeuratFiltered,
                           normalization.method = "LogNormalize",
                           scale.factor = 10000)

SeuratFiltered <- FindVariableFeatures(object = SeuratFiltered,
                                  selection.method = "vst",
                                  nfeatures = 2000)

all.genes <- rownames(x = SeuratObj)

SeuratFiltered <- ScaleData(object = SeuratFiltered, features = all.genes)
SeuratFiltered <- RunPCA(object = SeuratFiltered,
                    features = VariableFeatures(object = SeuratFiltered))


SeuratFiltered <- FindNeighbors(object = SeuratFiltered, dims = 1:10)
SeuratFiltered <- FindClusters(object = SeuratFiltered,
                          resolution = c(0.4, 0.5, 0.6, 0.8))


```

### View Genes

ST_plot can be used to view gene expression an image. The normalised scaled data
from the scale_data slot of a seurat object is shown.

NOTE: if this plot is used with a SingleCellExperiment object expression data 
will be taken from the logcounts slot.


```{r, genePlot, fig.height = 8, fig.width = 8, fig.align = "center", warning= F, message = F}


gene = "Nrgn"
ST_plot(Object = SeuratFiltered, Grob = image, 
        PlotType = "Gene", 
        Gene = gene)


```


### View Clusters

It can also be useful to look at how different clustering resolutions can be 
viewed match up with the image. Again this can done using the ST_plot. The 
ClusterRes argument should match a name in the meta.data to plot. 

NOTE: if this plot is used with a SingleCellExperiment object ClusterRes should match a column name within the colData.

```{r, clusterPlot, fig.height = 8, fig.width = 8, fig.align = "center", warning= F, message = F}

ST_plot(Object = SeuratFiltered, 
        Grob = image, 
        PlotType = "Cluster", 
        ClusterRes = "RNA_snn_res.0.6"
        )

```

## ShinySpaniel

## Using the RunShinySpaniel function in R

Spaniel includes a shiny app designed to share preprocessed data. The app is 
included as the runShinySpaniel function. 

### Mark Cluster Columns

The shiny Spaniel app has the option of visualising any cluster data on a plot.
First the columns containing the different clustering solutions need to be 
marked in the meta.dat with the prefix "cluster_"  so that Spaniel knows where 
to look for them. 

```{r, markClusters}
SeuratFiltered = markClusterCol(SeuratFiltered, "res")
```

### Save data object

```{r, eval = F}
saveRDS(SeuratFiltered, "data.rds")
```

An example of processed data set can be found:

```{r, eval = F}
file.path(system.file(package = "Spaniel"), "data/SeuratData.rds" )
```

### Save image object

```{r, eval = F}
saveRDS(SeuratFiltered, "image.rds")

```
An example of processed image file  can be found:

```{r, eval = F}
file.path(system.file(package = "Spaniel"), "data/image.rds" )
```

### Run shinySpaniel!
```{r, markclusterCols, eval = F}
runShinySpaniel()

```

## Deploying ShinySpaniel
For sharing analysis it is also possible to deploy the ShinySpaniel app to 
Shiny.io or any other server running R Server. The standalone app is located:

```{r, eval = F}
SpanielApp = file.path(system.file(package = "Spaniel"), "ShinySpaniel" )
```

You can sign up for a shinyapps.io account:

https://www.shinyapps.io/

Sign into your account and follow steps 1 and and 2 in the getting started 
instructions to authorise your account in R. 

Then run the following code:


```{r, eval = F}
library(rsconnect)
rsconnect::deployApp(SpanielApp, account = "spaniel")

```






